### PR TITLE
Issue #8748: Create new check RecordComponentNumberCheck

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #8748 : Create new check RecordComponentNumberCheck

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/e3988f8092d919b5cbe70f5359c4d9e8/raw/a0f6f5860f8025c19868061dee6e0e40960b992f/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/nmancus1/497dbdc31fe7a09cd960fc255972fb65/raw/c68be9ca06b61e1f279705c8b915386e701803c0/RecordComponentNumberCheckBase.xml

Diff Regression patch config: https://gist.githubusercontent.com/nmancus1/e40c93aa8c632af54f63a8c3347821da/raw/2734393e077ea5f2deb69f3570e0200452ab31fe/RecordComponentNumberCheckPatch.xml

Contribution PR: https://github.com/checkstyle/contribution/pull/500

Link to site: https://nmancus1.github.io/site_recordcomponentnumber/

~~First check regression report, before modifying ScopeUtil.java: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/full-records-RecordComponentNumberCheck_2020133606/reports/diff/index.html~~

Second check regression report, after merging #8598: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/full-records-RecordComponentNumberCheck_2020114210/reports/diff/index.html

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
Github, generate report

**Comment by pbludov** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/full-records-RecordComponentNumberCheck_2020172402/reports/diff/index.html

@nmancus1 The regressions shows NPE in `ScopeUtil.getScopeFromMods`. This seems to be related to #8598

**Comment by nmancus1** [comment]
GitHub, generate report

**Comment by nmancus1** [comment]
@pbludov I failed to run the report again, please see original comment or above bot comment for updated report.  NPE is gone.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+     */
+    private static void countComponents(ComponentCounter counter, DetailAST ast) {
+        final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+        final DetailAST modifiers =
+                ast.findFirstToken(TokenTypes.MODIFIERS);
+
+        counter.setScope(ScopeUtil.getScopeFromMods(modifiers));
+
+        DetailAST componentToCount =
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:253]: This can be done using `TokenUtils::forEachChild`

File: src/xdocs/checks.xml [hunk 2]
```diff
@@ -729,6 +729,10 @@
               Checks that pattern variable names conform to a specified pattern.
             </td>
           </tr>
+          <tr>
+            <td><a href="config_sizes.html#RecordComponentNumber">RecordComponentNumber</a></td>
+            <td> Checks the number of record components in the header of a record definition.</td>
```
Comment (pbludov) [inline src/xdocs/checks.xml:734]: Extra space after `<td>`

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java [hunk 3]
```diff
@@ -44,7 +44,7 @@ public static Scope getScopeFromMods(DetailAST aMods) {
         Scope returnValue = Scope.PACKAGE;
         for (DetailAST token = aMods.getFirstChild(); token != null
                 && returnValue == Scope.PACKAGE;
-                token = token.getNextSibling()) {
+             token = token.getNextSibling()) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:47]: Now the indent is a single space. Please revert or reformat this expression completely, may be
```java
for (DetailAST token = aMods.getFirstChild();
        token != null && returnValue == Scope.PACKAGE;
        token = token.getNextSibling()) {
    if ("public".equals(token.getText())) {
```
?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+        /** The scope of the record definition. */
+        private Scope scope;
+
+        /** Tracks the total. */
+        private int total;
+
+        /**
+         * Increments to counter by one.
+         *
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:322]: Extra blank line in the end of the comment

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    private void checkMax(int max, int value, String msg, DetailAST ast) {
+        if (max < value) {
+            log(ast, msg, value, max);
+        }
+    }
+
+    /**
+     * Counter class used to track the scope of the record definition and the number of
+     * components in the record component list.
+     *
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:310]: Extra blank line in the end of the comment

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        checkCounter(counter, ast);
+    }
+
+    /**
+     * Determine the visibility modifier, then iterate through the record component list
+     * and increment the counter.
+     *
+     * @param counter the ComponentCounter instance for this check
+     * @param ast the ast to check
+     *
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:243]: Extra blank line in the end of the comment

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:253]: Can you give me an example?  `forEachChild()` performs an action on each child, I'm not sure how that applies here.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 6 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 5 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 4 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java (hunk 3 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java]: Fixed.

File: src/xdocs/checks.xml (hunk 2 above)
Comment (nmancus1) [reply src/xdocs/checks.xml]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 1 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:252]:
```java
        TokenUtil.forEachChild(recordComponents,
                TokenTypes.RECORD_COMPONENT_DEF, node -> counter.increment());
```

What your code actually does:
1. For each child from the first child find one with type==TokenTypes.RECORD_COMPONENT_DEF.
2. Double check the type. Then increment the counter.
3. loop over the rest of children
It can be simplified to
```java
        DetailAST componentToCount = recordComponents.getFirstChild();
        while (componentToCount != null) {
            if (componentToCount.getType() == TokenTypes.RECORD_COMPONENT_DEF) {
                counter.increment();
            }
            componentToCount = componentToCount.getNextSibling();
        }
```
... and this is exactly what `TokenUtils::forEachChild` does:
1. loop for each child
2. compare type
3. if match, invoke callback

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Thanks a lot, @pbludov .  It is done.

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
Github, generate report

File: src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.sizes.recordcomponentnumber;
+
+/* Config:
+ * maxPublic = 10
```
Comment (romani) [inline src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java:5]: maxComponents

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        AccessModifierOption.PACKAGE,
+        AccessModifierOption.PRIVATE,
+    };
+
+    /**
+     * Setter to specify the maximum number of components allowed in record definition headers.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxComponents(int value) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:149]: Other properties like this just named `max`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    /**
+     * Check the counter and report violations.
+     *
+     * @param counter the component counter to check
+     * @param ast to report violations against.
+     */
+    private void checkCounter(ComponentCounter counter, DetailAST ast) {
+        if (matchAccessModifiers(counter.getAccessModifierOption())
+                && counter.getTotal() > maxComponents) {
+            log(ast, MSG_KEY, counter.getTotal(), maxComponents);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:222]: Please move this to visitToken method, we should see logic in that method

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Instantiate new component counter
+        final ComponentCounter counter = new ComponentCounter();
+
+        // Count record components in this record definition
+        countComponents(counter, ast);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:189]: Return a value from this method

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 8 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:149]: So, `setMax()`?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 10 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:189]: I've returned a ComponentCounter, so that we can also track the access modifier associated with the components we are counting.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 9 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:222]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 8 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:149]: Done.

File: src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java (hunk 7 above)
Comment (nmancus1) [reply src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java:5]: maxComponents -> max

File: src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber1.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.sizes.recordcomponentnumber;
+
+/* Config:
+ * maxPublic = 8
```
Comment (romani) [inline src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber1.java:5]: ?

File: src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java (hunk 7 above)
Comment (romani) [reply src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber2.java:5]: Not done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Count record components in this record definition
+        final ComponentCounter counter = countComponents(ast);
+
+        // Check counter and report violations
+        if (matchAccessModifiers(counter.getAccessModifierOption())
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:189]: Do not even count if modifier is not matching

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    private boolean matchAccessModifiers(final AccessModifierOption accessModifier) {
+        return Arrays.stream(accessModifiers)
+                .anyMatch(modifier -> modifier == accessModifier);
+    }
+
+    /**
+     * Counter class used to track the access modifier of the record definition and the number of
+     * components in the record component list.
+     */
+    private static class ComponentCounter {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:244]: Please remove this class, to just count you need single variable.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 12 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Much better, done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 13 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Removed, thank you.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 10 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: now returning an `int`, fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 9 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Improved further.

File: src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber1.java (hunk 11 above)
Comment (nmancus1) [reply src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentnumber/SuppressionXpathRecordComponentNumber1.java]: Got sleep, and fixed.

**Comment by nmancus1** [comment]
Direct link to RecordComponentNumber site: https://nmancus1.github.io/site_recordcomponentnumber/config_sizes.html#RecordComponentNumber

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks the number of record components in the header of a record definition.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:35]: it would be awesome to find some link in jdk that explain what is "header"

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+ * Type is {@code com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption[]}.
+ * Default value is {@code public, protected, package, private}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentNumber"/&gt;
+ * </pre>
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:56]: please add example of code for default config to should behavior on default config

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    public int[] getAcceptableTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:180]: reuse `getAcceptableTokens()`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Check access modifier of record definition
+        final DetailAST modifiers =
+            ast.findFirstToken(TokenTypes.MODIFIERS);
+        final AccessModifierOption accessModifier =
+            CheckUtil.getAccessModifierFromModifiersToken(modifiers);
+
+        // If access modifier of record definition matches access modifier options,
+        // count components
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:193]: remove comment, code is self descriptive

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+        if (matchAccessModifiers(accessModifier)) {
+            final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+            final int componentCount = countComponents(recordComponents);
+
+            // Check counter and report violations
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:199]: removoe comment

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Check access modifier of record definition
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:186]: remvoe coment

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java [hunk 20]
```diff
@@ -42,9 +42,9 @@ private ScopeUtil() {
     public static Scope getScopeFromMods(DetailAST aMods) {
         // default scope
         Scope returnValue = Scope.PACKAGE;
-        for (DetailAST token = aMods.getFirstChild(); token != null
-                && returnValue == Scope.PACKAGE;
-                token = token.getNextSibling()) {
+        for (DetailAST token = aMods.getFirstChild();
+                 token != null && returnValue == Scope.PACKAGE;
+                 token = token.getNextSibling()) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:47]: if this is formattign refactoring , please move to  separate commit.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/sizes/messages_tr.properties [hunk 21]
```diff
@@ -1,19 +1,14 @@
...
-
-maxParam = {0,number,integer} değerinden daha fazla parametre mevcut {1,number,integer} değer bulundu).
-
-too.many.methods          = Toplam metot sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
-too.many.packageMethods   = Toplam ''package'' erişimli metot sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
-too.many.privateMethods   = Toplam ''private'' metot sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
-too.many.protectedMethods = Toplam ''protected'' metot sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
-too.many.publicMethods    = Toplam ''public'' metot sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
+executableStatementCount=Yürütülebilir ifade sayısı {0,number,integer} (maksimum izin verilen değer {1,number,integer}).
+maxLen.anonInner=İsimsiz iç sınıf uzunluğu  {0,number,integer} satır (maksimum izin verilen değer {1,number,integer}).
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/sizes/messages_tr.properties:3]: re- formatting should be done in separate commit.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+import java.awt.Shape;
+import java.util.ArrayDeque;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+import org.w3c.dom.Node;
+
+/* Config:
+ *
+ * public "max" = 5
```
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java:14]: lets avoid this.
in future we will load config from this comment, this approach will complicate everything.

single instance of Check should be used for file.

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java:14]: Can I just remove this test then?  The point of this one was to prove that we can use multiple configs for different access modifiers.

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java:14]: yes, just  remove test

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 15 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:58]: Done.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java (hunk 22 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/recordcomponentnumber/InputRecordComponentNumberDifferentMaxForScope.java]: Removed and added  a new one, set to only "private" access modifier.  This was required for test coverage.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/sizes/messages_tr.properties (hunk 21 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/sizes/messages_tr.properties]: Fixed, it was automatic formatting by Intellij that I missed.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java (hunk 20 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java]: Reverted.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 19 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Removed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 18 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Removed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 17 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Removed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 15 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:58]: Added.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 14 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Linked.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks the number of record components in the header of a record definition.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code max} - Specify the maximum number of components allowed
+ * in record definition headers.
```
[the remark this answers is not in the document]
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Plural -> singular

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 14 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: please use http://cr.openjdk.java.net/~gbierman/jep359/jep359-20191031/specs/records-jls.html#jls-8.10.1
to point closely to where it is mentioned.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+ * </p>
+ * <ul>
+ * <li>
+ * {@code too.many.components}
+ * </li>
+ * </ul>
+ *
+ * @since 8.36
+ */
+@FileStatefulCheck
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:135]: it is stateless.

stateless means ready for  Multithread mode (our future)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 14 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: I agree, but it makes the line too long, or breaks the link in the javadoc.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 24 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:135]: Is this in reference to the AtomicInteger usage?

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 14 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed, I thought that Checkstyle would complain about line length > 100.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Check is skipping lines with `http` just for this reason

**Comment by romani** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks the number of record components in the
+ * <a href="http://cr.openjdk.java.net/~gbierman/jep359/jep359-20191031/specs/records-jls.html#jls-8.10.1">
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:36]: https://docs.oracle.com/javase/specs/jls/se14/preview/specs/records-jls.html#jls-8.10.1
?

**Review by pbludov — CHANGES_REQUESTED** [review]
item

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java (hunk 25 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:36]: https://github.com/checkstyle/checkstyle/pull/8595#discussion_r476471757

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java:36]: Your link will allow us to remove `gbierman` from `whitelist.words`,  I will change it.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java]: Fixed.

**Review by pbludov — APPROVED** [review]
